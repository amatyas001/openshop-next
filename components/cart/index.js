export * from './CartContent/CartContent';
export * from './CartItem/CartItem';
export * from './CartPanel/CartPanel';
