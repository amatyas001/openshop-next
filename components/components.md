Elements are composed by [Chakra-UI](https://chakra-ui.com) components and preserve the library features so [Style Props](https://chakra-ui.com/style-props) can be passed to them and accessibility remains optimized.

> _Style Props are a way to alter the style of a component by simply passing props to it. It helps to save time by providing helpful shorthand ways to style components._

For more information visit the [Chakra UI](https://chakra-ui.com) documentation.

Components are divided into two main category: _wrappers_ and _single responsibility_ elements. Wrapper components does not contain business logic, however there are special cases like _handling lifted states to keep sensitive data locally_ and not to populate them to the Redux store which is persisted on the machine.

**State Dependencies** and **Wrapped Components** are documented in each description to give brief information about the context of posible interactions of the actual component. _Examples_ prepared with proper state and may render extra controls to test different cases with ease.
