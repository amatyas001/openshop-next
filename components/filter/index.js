export * from './FilterButton/FilterButton';
export * from './FilterColor/FilterColor';
export * from './FilterPrice/FilterPrice';
export * from './FilterName/FilterName';
export * from './FilterPanel/FilterPanel';
export * from './FilterReset/FilterReset';
