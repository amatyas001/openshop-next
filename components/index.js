export * from './shared';
export * from './cart';
export * from './payment';
