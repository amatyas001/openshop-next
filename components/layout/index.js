export * from './Content/Content';
export * from './Footer/Footer';
export * from './ProductsTopBar/ProductsTopBar';
export * from './Navbar/Navbar';
export * from './ScrollTop/ScrollTop';
