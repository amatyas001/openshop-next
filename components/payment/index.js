export * from './PaymentCancelled/PaymentCancelled';
export * from './PaymentConfirm/PaymentConfirm';
export * from './PaymentError/PaymentError';
export * from './PaymentForm/PaymentForm';
export * from './PaymentProgress/PaymentProgress';
export * from './PaymentReview/PaymentReview';
export * from './PaymentSuccess/PaymentSuccess';
