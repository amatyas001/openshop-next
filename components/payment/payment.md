Each stage of the payment process is presented as child component under the `/checkout` page and assumes that `cart` is not empty. If `cart` is empty, than `/checkout` page will be redirect the user to the homepage.

Chekcout can be initiated from the [Cart](#section-cart) by clicking the _Continue_ button in the [Cart Panel](#cartpanel). This will load [PaymentReview](#paymentreview) stage and payment process becomes available under the `/checkout` page.

Components are interacting with the `payment` object and displayed according to `payment.status`. Payment data stored as various properties of the `payment` object. Detailed information can be found in each component documentation _(See: "State Dependencies")_.

Payment uses [Stripe](https://stripe.com) to handle secure customisable transactions. Requests to Stripe API are tokenized and _idempotent_.

On the case when user leave the payment process or close the browser tab, [Redux Persit](https://github.com/rt2zz/redux-persist) will store the current stage on the machine so process can be continued from where it was stopped.

**Stages**

1. **_Review_**

   - **[PaymentReview](#paymentreview)** is the initial stage. The user has options to go back browsing or continue the process.

2. **_Details_**

   1. **[PaymentForm](#paymentform)** is presented and on submit it will create an [Intent](https://stripe.com/docs/api/payment_intents/create) with [Client Secret](https://stripe.com/docs/payments/accept-a-payment).
   2. **[PaymentConfirm](#paymentconfirm)** using the _secret_ from [PaymentForm](#paymentform), asks the user to enter card details and confirm the payment.

3. **_Finished_**

   - **[PaymentSuccess](#paymentsuccess)** is shown when transaction was successful.
   - **[PaymentCancelled](#paymentcancelled)** is shown when user cancelled the payment process.
   - **[PaymentError](#paymenterror)** is shown when somthing bad happened during the transaction.

[PaymentProgress](#paymentprogress) is always visible on the `/checkout` page representing the payment process stages.

**State Dependencies** and **Wrapped Components** are documented in each description to give brief information about the context of posible interactions of the actual component. _Examples_ prepared with proper state and may render extra controls to test different cases with ease.
