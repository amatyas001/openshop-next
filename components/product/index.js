export * from './ProductCard/ProductCard';
export * from './ProductControls/ProductControls';
export * from './ProductControlsColor/ProductControlsColor';
export * from './ProductControlsAmount/ProductControlsAmount';
export * from './ProductControlsButton/ProductControlsButton';
export * from './ProductControlsSize/ProductControlsSize';
