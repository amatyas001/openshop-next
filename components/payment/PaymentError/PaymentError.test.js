import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import createStore from 'redux-mock-store';
import { paymentReset } from '@app/redux/action/payment';
import { PaymentError } from '@app/components';

let mockStore = createStore([]);

describe('<PaymentError />', () => {
  let tree,
    store = mockStore({
      payment: { status: 'error', error: { message: 'mock_error' } },
    });

  beforeAll(() => {
    act(() => {
      tree = create(
        <Provider store={store}>
          <PaymentError />
        </Provider>
      );
    });
  });

  it('should render without props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render error message', () => {
    expect(
      tree.root.findByProps({
        'data-testid': 'error-message',
      }).props.children
    ).toContain(store.getState().payment.error.message);
  });

  it('should render retry button', () => {
    expect(
      tree.root.findByProps({ 'data-testid': 'error-button-retry' })
    ).toBeTruthy();
  });

  it('should push reset state', async () => {
    await act(async () => {
      await tree.root
        .findByProps({ 'data-testid': 'error-button-retry' })
        .props.onClick();
    });

    expect(store.getActions()).toEqual([paymentReset()]);
  });
});
