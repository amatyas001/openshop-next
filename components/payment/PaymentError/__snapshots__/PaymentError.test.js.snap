// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaymentError /> should render with props 1`] = `
<div
  className="css-1mi3tt8"
>
  <p
    className="css-1po2p9s"
  >
    <strong>
      Error:
    </strong>
    <br />
    <em>
      mock_message
    </em>
  </p>
  <button
    className="css-187kagv"
    data-cy="button"
    data-testid="error-button-retry"
    href="#0"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    retry
  </button>
</div>
`;

exports[`<PaymentError /> should render without props 1`] = `
<div
  className="css-1mi3tt8"
>
  <p
    className="css-1po2p9s"
  >
    <strong>
      Error:
    </strong>
    <br />
    <em />
  </p>
  <button
    className="css-187kagv"
    data-cy="button"
    data-testid="error-button-retry"
    href="#0"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    retry
  </button>
</div>
`;
